### ISSUE TITLE
openstack packer provisioner with volume fails with error on ceph backend

### ISSUE THREAD

**Issue by nicolasochem** [hashicorp/packer#8009]
There appears to be a regression.

I am using packer openstack builder with the following configuration:

```
  "builders": [
    {
    "type": "openstack",
    "domain_name": "Default",
    "networks": ["{{ user `openstack_provider_network_id`}}"],
    "region": "regionOne",
    "ssh_username": "centos",
    "security_groups": "packer_provisioner",
    "image_name": "{{ user `openstack_image_name`}}",
    "source_image_name": "{{ user `openstack_source_image_name`}}",
    "flavor": "packer",
    "use_blockstorage_volume": "true",
    "volume_size": 56,
    "image_disk_format": "raw"
    }
  ]
```

After successful provisioning on an openstack vm, packer attempts to convert the volume to an image. It fails. Observed behaviour with packer 1.4.3

```
==> openstack: Stopping server: a84e213c-7c5a-4b82-a54f-3c641d084d39 ...
    openstack: Waiting for server to stop: a84e213c-7c5a-4b82-a54f-3c641d084d39 ...
==> openstack: Detaching volume packer_5d5acea6-c511-ad7a-3649-145419dad7e1 (volume id: 4be153c6-0333-42a5-aa5e-4803703b06c4)
==> openstack: Waiting for volume packer_5d5acea6-c511-ad7a-3649-145419dad7e1 (volume id: 4be153c6-0333-42a5-aa5e-4803703b06c4) to become available...
==> openstack: Creating the image: redacted
==> openstack: Error setting image metadata: Bad request with: [POST https://redacted/v3/6e95f572ba4449fba97791427d25fea4/volumes/4be153c6-0333-42a5-aa5e-4803703b06c4/action], error message: {"badRequest": {"message": "Invalid volume: Updat
ing volume metadata is not allowed for volumes in uploading status.", "code": 400}}
==> openstack: Terminating the source server: a84e213c-7c5a-4b82-a54f-3c641d084d39 ...
==> openstack: Deleting volume: 4be153c6-0333-42a5-aa5e-4803703b06c4 ...
==> openstack: Deleting temporary keypair: packer_5d5acea6-8871-8905-ae16-dc92bd0b7d7e ...
Build 'openstack' errored: Error setting image metadata: Bad request with: [POST https://redacted/v3/6e95f572ba4449fba97791427d25fea4/volumes/4be153c6-0333-42a5-aa5e-4803703b06c4/action], error message: {"badRequest": {"message": "Invalid v
olume: Updating volume metadata is not allowed for volumes in uploading status.", "code": 400}}
```

Expected behaviour (observed with packer 1.3.5):

```==> openstack: Stopping server: fe32f84d-db19-4c26-b840-9eed7922d95a ...
    openstack: Waiting for server to stop: fe32f84d-db19-4c26-b840-9eed7922d95a ...
==> openstack: Detaching volume packer_5d5ad655-cc33-a139-42cf-13871c552be0 (volume id: fd98a29d-5baa-4be3-8865-f38cb128bb61)
==> openstack: Waiting for volume packer_5d5ad655-cc33-a139-42cf-13871c552be0 (volume id: fd98a29d-5baa-4be3-8865-f38cb128bb61) to become available...
==> openstack: Creating the image: redacted
    openstack: Image: 2c52db5e-d94b-4b85-b554-2d10695ae656
==> openstack: Waiting for image redacted (image id: 2c52db5e-d94b-4b85-b554-2d10695ae656) to become ready...
==> openstack: Terminating the source server: fe32f84d-db19-4c26-b840-9eed7922d95a ...
==> openstack: Deleting volume: fd98a29d-5baa-4be3-8865-f38cb128bb61 ...
==> openstack: Deleting temporary keypair: packer_5d5ad655-3417-81ec-9401-5a80d2cfb29b ...
```

I am using ceph backend for block storage. Perhaps the recent changes in this area have not been tested against ceph.

**Comment by SwampDragons** [comment]
@psinghal20  can you take a look at this?

**Comment by sdemura** [comment]
Seeing this as well with packer 1.4.3

```
[Step 2/2] ==> some-region: Detaching volume packer_5d5bcc3e-569f-73fb-22fd-b5efa6aeed17 (volume id: 97cb5bbf-9fce-41bc-9107-245dac837976)
[Step 2/2] ==> some-region: Waiting for volume packer_5d5bcc3e-569f-73fb-22fd-b5efa6aeed17 (volume id: 97cb5bbf-9fce-41bc-9107-245dac837976) to become available...
[Step 2/2] ==> some-region: Creating the image: du-base-image-1566297149
[Step 2/2] ==> some-region: Error setting image metadata: Bad request with: [POST https://someplace.com/cinder/v3/b9815c903d6b4313b3be8f5915bfdf0d/volumes/97cb5bbf-9fce-41bc-9107-245dac837976/action], error message: {"badRequest": {"message": "Invalid volume: Updating volume metadata is not allowed for volumes in uploading status.", "code": 400}}
[Step 2/2] ==> some-region: Deleted temporary floating IP 'a824d913-a4c2-4c11-bd12-7a300754590a' (131.153.255.150)
[Step 2/2] ==> some-region: Terminating the source server: 550fe3ab-615e-4b7a-811c-f7df63b7d651 ...
[Step 2/2] ==> some-region: Deleting volume: 97cb5bbf-9fce-41bc-9107-245dac837976 ...
[Step 2/2] ==> some-region: Deleting temporary keypair: packer_5d5bcc3e-f2af-a283-8055-3eb8e0a0fbdf ...
[Step 2/2] Build 'some-region' errored: Error setting image metadata: Bad request with: [POST https://someplace.com/cinder/v3/b9815c903d6b4313b3be8f5915bfdf0d/volumes/97cb5bbf-9fce-41bc-9107-245dac837976/action], error message: {"badRequest": {"message": "Invalid volume: Updating volume metadata is not allowed for volumes in uploading status.", "code": 400}}
```

**Comment by psinghal20** [comment]
@SwampDragons The openstack upload API returns response code `202- Accepted: Request is accepted, but processing may take some time` [reference](https://docs.openstack.org/api-ref/block-storage/v3/?expanded=show-a-volume-s-details-detail,upload-volume-to-image-detail#upload-volume-to-image). Since the image is still being uploaded, we can't set the image metadata for that volume. Maybe we can block for image to be uploaded and then proceed with setting metadata.

**Comment by SwampDragons** [comment]
@psinghal20 that seems appropriate. We already wait for the image to be added in this step -- see WaitForImage below; we can just move the metadata addition until after this wait. I'd also suggest we skip the call to add metadata unless a user has provided ImageMetadata to add.

**Comment by nicolasochem** [comment]
thanks @psinghal20 !

**Comment by v0lumehi** [comment]
Thanks a lot for reporting this @nicolasochem. Had the same issue and it was driving me mad. Also thanks for fixing @psinghal20. I saw that the fix will be rolled out with version 1.4.4. Using the nightly packer build fixes it for me though

**Comment by v0lumehi** [comment]
It appears that the issue still occurs in the nightly packer build:
`$ packer --version
1.4.4
`
STDERR output:
`
Build 'openstack' errored: Error setting image metadata: Bad request with: [POST https://private.jambit.cloud:8776/v3/2b1777b0d2ec4263a6f0a63a0f313670/volumes/989b19ea-3b34-404e-aeb1-c666e8bde498/action], error message: {"badRequest": {"message": "Invalid volume: Updating volume metadata is not allowed for volumes in uploading status.", "code": 400}}
`
